server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -176,6 +176,16 @@
     private final KeyedLock<Long> noOpKeyedLock = new KeyedLock<>();
     private final AtomicBoolean shouldPeriodicallyFlushAfterBigMerge = new AtomicBoolean(false);
 
+    /**
+     * If multiple writes passed {@link InternalEngine#testReserveDocs(Operation, int)} (int)} but haven't adjusted
+     * {@link IndexWriter#getPendingNumDocs()} yet, then IndexWriter can fail with too many documents. In this case,
+     * we have to fail the engine because we already generated sequence numbers for write operations; otherwise we
+     * will have gaps in sequence numbers. To avoid this (best-effort), we use a lower limit when trying to reserve
+     * documents in InternalEngine.
+     */
+    static final int DEFAULT_MAX_DOCS = IndexWriter.MAX_DOCS - 1000;
```

Comment:
I'm wondering why we use a margin of 1000 docs here. IIUC the index writer will only under-count the total by at most one doc per write thread?

OTOH if there is more severe under-counting under concurrent load then is 1000 enough?

Reply:
> IIUC the index writer will only under-count the total by at most one doc per write thread?

Yes, but IW can under-count more than one with nested documents.

> OTOH if there is more severe under-counting under concurrent load then is 1000 enough?

I think 1000 is enough because all writes should execute using the WRITE thread pool, which has no more threads than the number of cores. Do you have a suggestion for this?

Reply:
Ah darn it yes, nested docs strike again. In which case I don't think 1000 is enough for all users. Can we instead acquire a proper reservation for the right number of docs in `testReserveDocs` (e.g. using a `Semaphore`) and then release them once they're tracked by the `IndexWriter`?

Reply:
Adjusted in https://github.com/elastic/elasticsearch/pull/63273/commits/efa8145303e98c3a1b9f4886691a5cfbd5c548b4